src/pilot/assign_role.py

```diff
@@ -846,7 +846,11 @@ def assign_role(ip_mac_service_tag, node_uuid, role_index,
     if role_index.index:
         role = "node:{}-{}".format(flavor, role_index.index)
 
-    value = "{},{},boot_mode:uefi,boot_option:local".format(role, node.properties['capabilities'])
+    if 'capabilities' in node.properties:
```

Comment:
`capabilities` is available if in-band introspection (IB) or out-of-band (OOB) inspection is performed before `assign_role.py` is executed, This `if` `else` condition will make this work in a controlled manner when in-band (IB) introspection or out-of-band (OOB) inspection has not been performed before `assign_role.py` is run and add `boot_mode` and `boot_option` to the node properties.

Reply:
clarified requirements:

If there are no capabilities then add capabilities with role, boot_mode:uefi and boot_option: local (as in line 857 already)
If there are capabilities, then
- add role
- for boot_mode and boot_option update or add it so that there are boot_mode:uefi, boot_option:local

Reply:
Please consider this approach: https://github.com/openstack/ironic/blob/349bf2d5bcb38b409fe01aa78cc281f016199efc/ironic/common/utils.py#L402.